### PR TITLE
REQ-627 Feature Branch

### PR DESCRIPTION
**Pull request by lindig**
This series of commits sets up the REQ-627 feature branch that prepares xapi to support SR-IOV for NVIDIA GPUs.

### CODE REVIEW

File: ocaml/xapi/vgpuops.ml [hunk 1]
```diff
@@ -102,7 +102,13 @@ let add_pcis_to_vm ~__context host vm pci =
...
-  let value = String.concat "," (List.map Pciops.to_string devs) in
+  let value =
+    Db.VM.get_other_config ~__context ~self:vm
+    |> (fun assoc -> try List.assoc Xapi_globs.vgpu_pci assoc with Not_found -> "")
+    |> ( function (* add to existing entry *)
+       | ""  -> String.concat "," (List.map Pciops.to_string devs)
+       | str -> String.concat "," (str :: List.map Pciops.to_string devs)
+       ) in
   debug "Adding PCIs to a VM with 'other config': %s" value;
   Db.VM.add_to_other_config ~__context ~self:vm ~key:Xapi_globs.vgpu_pci ~value
```
Comment (robhoes) [inline ocaml/xapi/vgpuops.ml]: Does this work now? Do you not need to remove first and then re-add?

File: ocaml/xapi/xapi_xenops.ml [hunk 2]
```diff
@@ -657,6 +657,18 @@ module MD = struct
       fn = 0;
     }
 
+  (** Return the virtual function (VF) for a VGPU operated in SR-IOV
+   * mode, or None otherwise
+   *)
+  let sriov_vf ~__context vgpu =
+    match vgpu.Db_actions.vGPU_pci with
```
Comment (robhoes) [inline ocaml/xapi/xapi_xenops.ml]: Where are you setting this field? So far, the PCI address only goes into the `other-config` field.

File: ocaml/xapi/vgpuops.ml (hunk 1 above)
Comment (lindig) [reply ocaml/xapi/vgpuops.ml]: Is `Db.VM.add_to_other_config` overwriting the existing entry (key) with the new value? The new value is constructed by adding the new value to the existing value.

File: ocaml/xapi/xapi_xenops.ml (hunk 2 above)
Comment (lindig) [reply ocaml/xapi/xapi_xenops.ml]: I agree, this does not work as the value is never set.

**Comment by lindig** [comment]
My current plan:

* Raise error if Nvidia script can't be found
* Make path to Nvidia script configurable
* Use vGPU PCI address in datamodel and avoid using other-config key on VM

**Comment by lindig** [comment]
I have addressed the comments, please take a fresh look.

File: ocaml/idl/datamodel.ml [hunk 3]
```diff
@@ -4780,6 +4780,11 @@ module VGPU = struct
         field ~qualifier:DynamicRO ~ty:(Ref _pgpu) ~lifecycle:[Published, rel_dundee, ""] "scheduled_to_be_resident_on" "The PGPU on which this VGPU is scheduled to run" ~default_value:(Some (VRef null_ref));
         field ~qualifier:DynamicRO ~ty:(Map (String,String)) ~lifecycle:[Published, rel_inverness, ""] ~default_value:(Some (VMap [])) "compatibility_metadata" "VGPU metadata to determine whether a VGPU can migrate between two PGPUs";
         field ~qualifier:RW ~ty:String ~lifecycle:[Published, rel_quebec, ""] ~default_value:(Some (VString "")) "extra_args" "Extra arguments for vGPU and passed to demu";
+        field ~qualifier:DynamicRO
+          ~ty:(Ref _pci)
+          ~lifecycle:[Published, rel_quebec, ""]
+          ~default_value:(Some (VRef null_ref))
+          "pci" "PCI device";
```
Comment (robhoes) [inline ocaml/idl/datamodel.ml]: For consistency with other API fields, we should capitalise this name: `PCI`.
Also, please explain in the doc string that this is the PCI device that is passed through to the VM for this VGPU, either a full device pass-through, or an SR-IOV virtual function.

**Comment by robhoes** [comment]
The commit message of d46cbde seems to be truncated. Other than that, it all looks good to me.

**Comment by lindig** [comment]
I have addressed the comments and rebased. I will address the build issues and would merge this once they are resolved.

**Pull request merged**